### PR TITLE
[vNext CCB] Tokenizing Command Bar

### PR DESCRIPTION
**Pull request by sophialee0416**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

Replacing CCB specs with tokens.

### Verification

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| Simulator Screen Shot - iPhone 11 Pro - 2021-09-08 at 17 12 06 | Simulator Screen Shot - iPhone 11 Pro - 2021-09-08 at 17 09 20 |
| Simulator Screen Shot - iPhone 11 Pro - 2021-09-08 at 17 11 38 | Simulator Screen Shot - iPhone 11 Pro - 2021-09-17 at 14 29 46 |

**Theming:**

https://user-images.githubusercontent.com/22566866/133855210-f53a99d0-570c-49c8-b0ae-a79241c85c61.mov

**UIPointerInteraction Testing**

https://user-images.githubusercontent.com/22566866/134067023-c1771feb-eece-4cab-9553-d15b58d81c98.mov

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [ ] Different resolutions (1x, 2x, 3x)
- [ ] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/705)

### CODE REVIEW

**Review by mischreiber — COMMENTED** [review]
I tested swapping the color theme from blue to green in the test app, but the CCB did not update. Were you able to verify this scenario on your end?

File: ios/FluentUI/Command Bar/CommandBarTokens.swift [hunk 1]
```diff
@@ -0,0 +1,52 @@
+//
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//  Licensed under the MIT License.
+//
+
+import UIKit
+
+class MSFCommandBarTokens: MSFTokensBase, ObservableObject {
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBarTokens.swift:8]: This class does not need to be an ObservableObject as the CCB implementation is not SwiftUI.

File: ios/FluentUI/Command Bar/CommandBarTokens.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    @Published public var defaultBackgroundColor: UIColor!
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBarTokens.swift:9]: Same here for the @Published properties. We don't need these if we're not using it in a SwiftUI View.

File: ios/FluentUI/Command Bar/CommandBarTokens.swift [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    @objc open func didChangeAppearanceProxy() {
+        updateForCurrentTheme()
+    }
+
+    override func updateForCurrentTheme() {
+        let currentTheme = theme
+        let appearanceProxy = currentTheme.MSFCommandBarTokens
+
+        defaultBackgroundColor = appearanceProxy.backgroundColor.rest
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBarTokens.swift:41]: Why are you naming these properties "default" instead of just backgroundColor and iconColor?

File: ios/FluentUI/Command Bar/CommandBar.swift [hunk 4]
```diff
@@ -11,7 +11,7 @@ import UIKit
  Set the `delegate` property to determine whether a button can be selected and deselected, and listen to selection changes.
  */
 @objc(MSFCommandBar)
-open class CommandBar: UIView {
+open class CommandBar: UIView, FluentUIWindowProvider {
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBar.swift:14]: You need to override [didMoveToWindow()](https://developer.apple.com/documentation/uikit/uiview/1622527-didmovetowindow) to update the tokens as they're based on the current window.

File: tools/sgen/input/FluentUIStyle.yml [hunk 5]
```diff
@@ -311,6 +311,16 @@ MSFBorderedCardNudgeTokens extends MSFCardNudgeTokens:
     outlineColor: $Colors.Stroke.neutral1
     textColor: $Colors.Brand.shade20
 
+AP_MSFCommandBarTokens:
```
Comment (rdeassis) [inline tools/sgen/input/FluentUIStyle.yml:314]: We should change the Demo Controller description to "CommandBar (Vnext)" to communicate that this control is tokenized.

File: ios/FluentUI/Command Bar/CommandBarTokens.swift [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Published public var dismissBackgroundColor: UIColor!
+    @Published public var dismissIconColor: UIColor!
+    @Published public var groupBorderRadius: CGFloat!
+    @Published public var groupInterspace: CGFloat!
+    @Published public var iconSize: CGFloat!
+    @Published public var itemInterspace: CGFloat!
+
+    /// Notifies the drawer controller to refresh its UI to reflect its design token values
+    var themeDidUpdate: (() -> Void)?
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBarTokens.swift:23]: This is not being used anywhere.

File: ios/FluentUI/Command Bar/CommandBarTokens.swift [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    @Published public var pressedIconColor: UIColor!
+
+    @Published public var dismissBackgroundColor: UIColor!
+    @Published public var dismissIconColor: UIColor!
+    @Published public var groupBorderRadius: CGFloat!
+    @Published public var groupInterspace: CGFloat!
+    @Published public var iconSize: CGFloat!
+    @Published public var itemInterspace: CGFloat!
+
+    /// Notifies the drawer controller to refresh its UI to reflect its design token values
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBarTokens.swift:22]: "drawer controller"

File: ios/FluentUI/Command Bar/CommandBar.swift [hunk 8]
```diff
@@ -28,9 +28,14 @@ open class CommandBar: UIView {
...
     @objc public init(itemGroups: [CommandBarItemGroup], leadingItem: CommandBarItem? = nil, trailingItem: CommandBarItem? = nil) {
         self.itemGroups = itemGroups
+        self.leadingItem = leadingItem
+        self.trailingItem = trailingItem
 
         super.init(frame: .zero)
+        updateView()
+    }
 
+    private func updateView() {
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBar.swift:38]: nit: order of members in the file (private should be down below)

https://github.com/microsoft/swift-guide/blob/gh-pages/FileOrganization.md

**Review by rdeassis — CHANGES_REQUESTED** [review]

File: tools/sgen/input/FluentUIStyle.yml [hunk 9]
```diff
@@ -311,6 +311,16 @@ MSFBorderedCardNudgeTokens extends MSFCardNudgeTokens:
...
+  backgroundColor: [ rest: $Colors.Background.neutral4, pressed: $Colors.Background.brandRest ]
+  dismissBackgroundColor: $Colors.Background.neutral1
+  dismissIconColor: "FluentUIColor(light: $Colors.Foreground.neutral1, dark: $Colors.Foreground.neutral3)"
+  groupBorderRadius: $Border.radius.xLarge
+  groupInterspace: $Spacing.medium
+  iconColor: [ rest: $Colors.Foreground.neutral1, pressed: $Colors.Foreground.neutralInverted ]
```
Comment (rdeassis) [inline tools/sgen/input/FluentUIStyle.yml:320]: Don't we also need tokens for the disabled state?

**Comment by sophialee0416** [comment]
> I tested swapping the color theme from blue to green in the test app, but the CCB did not update. Were you able to verify this scenario on your end?

These should be fixed with the new iteration @mischreiber !

File: ios/FluentUI/Command Bar/CommandBarButton.swift [hunk 10]
```diff
@@ -14,6 +14,12 @@ class CommandBarButton: UIButton {
         }
     }
 
+//    override var isFocused: Bool {
```
Comment (harrieshin) [inline ios/FluentUI/Command Bar/CommandBarButton.swift:17]: remove?

File: ios/FluentUI/Command Bar/CommandBarTokens.swift [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    public var pressedBackgroundColor: UIColor!
+    public var pressedIconColor: UIColor!
+    
+    public var selectedBackgroundColor: UIColor!
+    public var selectedIconColor: UIColor!
+    
+    public var disabledBackgroundColor: UIColor!
+    public var disabledIconColor: UIColor!
+
+    public var dismissBackgroundColor: UIColor!
```
Comment (harrieshin) [inline ios/FluentUI/Command Bar/CommandBarTokens.swift:24]: what's dismiss background color for?

Comment (sophialee0416) [reply ios/FluentUI/Command Bar/CommandBarTokens.swift:24]: This is used for the keyboard dismiss!

Comment (harrieshin) [reply ios/FluentUI/Command Bar/CommandBarTokens.swift:24]: then there is clear mismatch in naming?

Comment (harrieshin) [reply ios/FluentUI/Command Bar/CommandBarTokens.swift:24]: the demo app is using for keyboard dismiss but the real button type(style) is fixed button isn't it?

Comment (sophialee0416) [reply ios/FluentUI/Command Bar/CommandBarTokens.swift:24]: I believe it means that if the button is not of persist selection type, then it will get the dismissed color properties, sorry if I am getting it confused

File: ios/FluentUI/Command Bar/CommandBarTokens.swift [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+//  Licensed under the MIT License.
+//
+
+import UIKit
+
+class MSFCommandBarTokens: MSFTokensBase {
+    public var backgroundColor: UIColor!
+    public var iconColor: UIColor!
+
+    public var hoverBackgroundColor: UIColor!
```
Comment (harrieshin) [inline ios/FluentUI/Command Bar/CommandBarTokens.swift:12]: where is hover color being used? keyboard focus? pointer color?

Comment (sophialee0416) [reply ios/FluentUI/Command Bar/CommandBarTokens.swift:12]: Updated with pointer usage

File: ios/FluentUI/Core/FluentUIFramework.swift [hunk 13]
```diff
@@ -17,7 +17,7 @@ public extension Colors {
     }
 
     internal struct NavigationBar {
-        static var background = UIColor(light: surfacePrimary, dark: gray900)
+        static var background: UIColor = surfacePrimary
```
Comment (harrieshin) [inline ios/FluentUI/Core/FluentUIFramework.swift:20]: this has bigger consequences on all of our navigation bar colors. especially on dark mode split mode cases with dark elevated concepts. should it be part of this change? shouldn't cab has its own background color?

File: tools/sgen/input/FluentUIStyle.yml [hunk 14]
```diff
@@ -311,6 +311,15 @@ MSFBorderedCardNudgeTokens extends MSFCardNudgeTokens:
     outlineColor: $Colors.Stroke.neutral1
     textColor: $Colors.Brand.shade20
 
+AP_MSFCommandBarTokens:
+  backgroundColor: [ rest: $Colors.Background.neutral4, hover: "FluentUIColor(light: $Colors.Background.neutral5, dark: $Colors.Stroke.neutral2)", pressed: "FluentUIColor(light: $Colors.Background.neutralDisabled, dark: $Colors.Background.neutral5)", selected: $Colors.Background.brandRest, disabled: $Colors.Stroke.neutral1 ]
+  dismissIconColor: "FluentUIColor(light: $Colors.Foreground.neutral1, dark: $Colors.Foreground.neutral3)"
```
Comment (harrieshin) [inline tools/sgen/input/FluentUIStyle.yml:316]: still reference on dismissIconColor

File: ios/FluentUI/Core/FluentUIFramework.swift [hunk 15]
```diff
@@ -29,6 +29,8 @@ public extension Colors {
 
     // Objective-C support
     @objc static var navigationBarBackground: UIColor { return NavigationBar.background }
+
+    @objc static var commandBarBackground: UIColor { return surfacePrimary }
```
Comment (harrieshin) [inline ios/FluentUI/Core/FluentUIFramework.swift:33]: why is this here? shouldn't it be part of token pipeline?

**Review by harrieshin — APPROVED** [review]

File: ios/FluentUI/Command Bar/CommandBarTokens.swift [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    override init() {
+        super.init()
+
+        self.themeAware = true
+
+        updateForCurrentTheme()
+    }
+
+    @objc open func didChangeAppearanceProxy() {
+        updateForCurrentTheme()
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBarTokens.swift:40]: I just noticed this issue where the primary brand color is not updated in the keyboard CCB:

https://user-images.githubusercontent.com/68076145/134248265-2340c2bf-725d-46ab-ae5f-652f51c6c66b.mov

File: ios/FluentUI/Command Bar/CommandBarButton.swift [hunk 17]
```diff
@@ -74,43 +82,90 @@ class CommandBarButton: UIButton {
...
 
+    private var commandBarTokens: MSFCommandBarTokens
+
     private let isPersistSelection: Bool
 
-    private var selectedTintColor: UIColor {
-        guard let window = window else {
-            return UIColor(light: Colors.communicationBlue, dark: .black)
-        }
+    private var hoverBackgroundColor: UIColor {
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBarButton.swift:89]: Do we need these computed properties? Why not just using their values directly?

File: ios/FluentUI/Command Bar/CommandBarTokens.swift (hunk 16 above)
Comment (mischreiber) [reply ios/FluentUI/Command Bar/CommandBarTokens.swift:40]: Good find! Isn't the keyboard in its own `UIWindow` instance? If the accessory bar is attached to that, then it may not have the same root window, and thus has a separate theme.

Comment (rdeassis) [reply ios/FluentUI/Command Bar/CommandBarTokens.swift:40]: @mischreiber yes, that's what I suspected too. And it does repro on the current CCB (before tokenization).
@sophialee0416 if this is the case, It'd be good to set the ColorProviding theme for the keyboard window so we have a good sample code for the client apps.

Comment (sophialee0416) [reply ios/FluentUI/Command Bar/CommandBarTokens.swift:40]: Resolved offline: will create a separate task for theming mismatch due to different UIWindows as a part of the theming initiative @mischreiber @rdeassis

File: tools/sgen/input/FluentUIStyle.yml [hunk 18]
```diff
@@ -311,6 +311,16 @@ MSFBorderedCardNudgeTokens extends MSFCardNudgeTokens:
     outlineColor: $Colors.Stroke.neutral1
     textColor: $Colors.Brand.shade20
 
+AP_MSFCommandBarTokens:
+  backgroundColor: [ rest: $Colors.Background.neutral4, hover: "FluentUIColor(light: $Colors.Background.neutral5, dark: $Colors.Stroke.neutral2)", pressed: "FluentUIColor(light: $Colors.Background.neutralDisabled, dark: $Colors.Background.neutral5)", selected: $Colors.Background.brandRest, disabled: $Colors.Stroke.neutral1 ]
+  barBackgroundColor: $Colors.Background.neutral1
```
Comment (rdeassis) [inline tools/sgen/input/FluentUIStyle.yml:316]: nit on the naming:
Shouldn't we have **barBackgroundColor** as just "**backgroundColor**"?

Then the ones related to the buttons, we can go with the "item" prefix?
- itemFixedIconColor
- itemIconColor
- itemIconSize
- itemInterspace

File: ios/FluentUI/Command Bar/CommandBar.swift [hunk 19]
```diff
@@ -103,9 +110,9 @@ open class CommandBar: UIView {
         scrollView.translatesAutoresizingMaskIntoConstraints = false
         scrollView.contentInset = UIEdgeInsets(
             top: 0,
-            left: leadingButton == nil ? CommandBar.insets.left : CommandBar.fixedButtonSpacing,
+            left: leadingButton == nil ? CommandBar.insets.left : commandBarTokens.itemInterspace,
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBar.swift:113]: nit: store "commandBarTokens.itemIterspace" in a constant (let) and use it in both of these lines (113 and 115).

Comment (rdeassis) [reply ios/FluentUI/Command Bar/CommandBar.swift:113]: Same for lines 179, 181, 193 and 195.

**Review by rdeassis — APPROVED** [review]

**Review by mischreiber — APPROVED** [review]

File: ios/FluentUI/Command Bar/CommandBar.swift [hunk 20]
```diff
@@ -38,12 +39,18 @@ open class CommandBar: UIView {
...
+        backgroundColor = commandBarTokens.backgroundColor
         translatesAutoresizingMaskIntoConstraints = false
 
         configureHierarchy()
         updateButtonsState()
     }
 
+    open override func didMoveToWindow() {
+        super.didMoveToWindow()
+        commandBarTokens.didChangeAppearanceProxy()
```
Comment (rdeassis) [inline ios/FluentUI/Command Bar/CommandBar.swift:51]: Hey @sophialee0416. While reviewing the tokenization of the NotificationView earlier today I made this comment that I missed in the CommandBar review:

Could you create a quick PR to update this call to updateForCurrentTheme() instead?
Thanks!

https://github.com/microsoft/fluentui-apple/pull/715#discussion_r718039047
